### ISSUE TITLE
Miss cache file problem on Jenkins concurrent install

### ISSUE THREAD

**Issue by yanzhiwei147** [CocoaPods/CocoaPods#6280]
# Report

## What did you do?

Run `pod update --verbose --no-repo-update` in Jenkins(3 concurrent jobs )

## What did you expect to happen?

Install all pod dependencies correctly and compile correctly(each pod  provide [framework](https://github.com/jverkoey/iOS-Framework),not source code.and we have a private cocoapod repo ).

## What happened instead?

```sh
# cache dir
➜  9.0.0-afaf5 tree
.
|-- LBSComponent.framework
|   |-- Headers -> Versions/Current/Headers
|   |-- LBSComponent -> Versions/Current/LBSComponent
|   `-- Versions
|       |-- A
|       |   |-- Headers
|       |   |   `-- LBSComponent.h
|       |   `-- LBSComponent
|       `-- Current -> A
`-- README.md

6 directories, 4 files
➜  9.0.0-afaf5 pwd
/Users/jenkins/Library/Caches/CocoaPods/Pods/Release/LBSComponent/9.0.0-afaf5

# job1 install log(install fail)
-> Installing LBSComponent (9.0.0)
  > Copying LBSComponent from `/Users/jenkins/Library/Caches/CocoaPods/Pods/Release/LBSComponent/9.0.0-afaf5` to `Pods/LBSComponent`

# job1 error(install fail, dir struct incorrect)
RuntimeError - unknown file type: /Users/jenkins/Library/Caches/CocoaPods/Pods/Release/QRCodeComponent/0.2.21-73dff/QRCodeComponent.framework/Versions/Current
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1375:in `copy'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:463:in `block in copy_entry'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1485:in `call'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1485:in `wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1488:in `block in wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1488:in `block in wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1488:in `block in wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1487:in `wrap_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:460:in `copy_entry'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:435:in `block in cp_r'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1558:in `block in fu_each_src_dest'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1574:in `fu_each_src_dest0'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:1556:in `fu_each_src_dest'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:434:in `cp_r'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/downloader.rb:54:in `block in download'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:142:in `message'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/downloader.rb:52:in `download'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/pod_source_installer.rb:120:in `download_source'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/pod_source_installer.rb:60:in `install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:393:in `install_source_of_pod'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:360:in `block (2 levels) in install_pod_sources'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:85:in `titled_section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:359:in `block in install_pod_sources'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:351:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:351:in `install_pod_sources'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:197:in `block in download_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:195:in `download_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:123:in `install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/update.rb:60:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:23:in `load'
/usr/local/bin/pod:23:in `<main>'

# job1 Pods/LBSComponent (copy fail)
➜  LBSComponent git:(4225282) ✗ tree
.
|-- LBSComponent.framework
|   `-- Versions
|       `-- Current -> A
`-- README.md

2 directories, 2 files
➜  LBSComponent git:(cccb027) ✗ pwd
/data/jenkins/release/workspace/fac-com-iOS_page1125011480302344208/Pods/LBSComponent

# job2 install log(install success) 
-> Installing LBSComponent (9.0.0)
  > Copying LBSComponent from `/Users/jenkins/Library/Caches/CocoaPods/Pods/Release/LBSComponent/9.0.0-afaf5` to `Pods/LBSComponent`

# job2 error(compile fail)
ld: framework not found LBSComponent
clang: error: linker command failed with exit code 1 (use -v to see invocation)

# job2 Pods/LBSComponent (copy success, dir struct incorrect)
➜  LBSComponent git:(4225282) ✗ tree
.
|-- LBSComponent.framework
|   `-- Versions
|       `-- Current -> A
`-- README.md

2 directories, 2 files
➜  LBSComponent git:(4225282) ✗ pwd
/data/jenkins/release/workspace/fac-com-iOS_maincomponentcom1452585173662/Pods/LBSComponent
```

### Stack

```
   CocoaPods : 1.1.1
        Ruby : ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin15]
    RubyGems : 2.4.6
        Host : Mac OS X 10.11.5 (15F34)
       Xcode : 8.0 (8A218a)
         Git : git version 2.8.4 (Apple Git-73)
Ruby lib dir : /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib
Repositories : master - https://github.com/CocoaPods/Specs.git @ dfd7f3ff449699ffb5add53e1c58cd8b3df5c04a
               sdp-cocoapods-spec - http://git.sdp.nd/cocoapods/spec.git @ b445fd7b4616bc2de3699f13a84f35963753ce4e
               sdp-mirrors-specs - http://git.sdp.nd/Mirrors/specs.git @ 90cb0aa7c75e4bcc2796138f679b0d2434d4db5d
               sdp_repo - http://git.sdp.nd/cocoapods/spec.git @ 576bd6aaabf1228ca09d593ff2125cb1467e943c
```

### Plugins

```
cocoapods-deintegrate : 1.0.1
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.1.1
cocoapods-try         : 1.1.0
```

**Comment by yanzhiwei147** [comment]
I found other problem,each job should be run `pod repo update private-repo`, it not support concurrent execute.

**Comment by benasher44** [comment]
Hi there! Many CocoaPods `pod` commands should support being run concurrently. Unfortunately, running `pod` commands that do `git` operations on shared spec repos may fail depending on how `git` handles it.  Are these concurrent jobs running `pod update --no-repo-update` on the same project?

**Comment by yanzhiwei147** [comment]
@benasher44 No,look my comment above,it run on two project.
```sh
#project1 pwd
➜  LBSComponent git:(4225282) ✗ pwd
/data/jenkins/release/workspace/fac-com-iOS_page1125011480302344208/Pods/LBSComponent

#project2 pwd
➜  LBSComponent git:(4225282) ✗ pwd
/data/jenkins/release/workspace/fac-com-iOS_maincomponentcom1452585173662/Pods/LBSComponent
```

**Comment by benasher44** [comment]
Hmm. Would you mind sharing your Podfile? There could be other `git` operations at play here. Thanks!

**Comment by yanzhiwei147** [comment]
@benasher44 Sorry, the Podfile's content not work for you. It provide by our Private Spec Repo.I think you can execute `pod update` concurrent, issue should reappear soon.

**Comment by yanzhiwei147** [comment]
Sorry,can we reappear the issues?

**Comment by benasher44** [comment]
@yanzhiwei147 is there any more information you can provide about the pods being installed? Maybe you could share the podspecs?

**Comment by yanzhiwei147** [comment]
@benasher44 Yes of course,what the information would you get?

**Comment by benasher44** [comment]
@yanzhiwei147 okay I'm looking at this some more, and it appears that we don't do anything to "lock" the cache directory. I'd say that this falls into the category of scm operations (fetching sources into a cache) that we don't guarantee to be safe for multiple processes to perform under the same user. The easiest way around this would be to just commit your Pods directory, since then no sources would need to be fetched, and `pod install` would be limited to setting up your local workspace/project.

Alternatively, we could lock the cache directory. The issue here is that `pod install` currently isn't good at being killed. Today, if you kill `pod install` your project/caches will be left in an unknown state. To support this, we'd have to be more diligent about ensuring that we clean up (undo the lock in this case) when `pod install` is killed.

**Comment by charlesdbs** [comment]
`def cached_spec(request)
        path = path_for_spec(request)
        path.file? && Specification.from_file(path)
      rescue JSON::ParserError
        nil
      end`
it seems that we just check path exists， so when multiple processes runs path mdkir,but copy not end, we can not install complete cache into Pods.

**Comment by yanzhiwei147** [comment]
@benasher44 Sorry,I forget make this PR. It can be resolved by use file lock.

**Comment by codiophile** [comment]
@dnkoutso @amorde This can be closed as well. Fixed by #10018. :)
